### ISSUE TITLE
docker_container: add `cleanup` state value

### ISSUE THREAD

**Issue by srstsavage** [ansible/ansible-modules-core#4088]
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

`docker_container` module
##### ANSIBLE VERSION

(Future version)
##### SUMMARY

A `cleanup` parameter should be added to `docker_container` to remove the container as soon as it's finished running. `docker run` supports this using `--rm`. This would be especially useful for containers run using `detach: no` to perform one time tasks.

Currently I'm having to hack around this by assigning a unique name to the container (appending `ansible_date_time.epoch` to a string) and deleting it immediately after. Using a static string as the container name isn't advised because parallel ansible executions could lead to a race condition.

@chouseknecht

**Comment by nitzmahone** [comment]
cc @chouseknecht

**Comment by PikachuEXE** [comment]
I think the option can be named `cleanup` to align with what the Docker doc called the option
https://docs.docker.com/engine/reference/run/#/clean-up-rm

**Comment by srstsavage** [comment]
@PikachuEXE makes sense to me, renamed proposed `remove` parameter to `cleanup`.

**Comment by denise-sanders** [comment]
The default value should be false, right? Because that is what people will expect? (I am looking at implementing this)

**Comment by srstsavage** [comment]
@deniseirvin Yeah, default should definitely be false. You could make the argument that defaulting to true when using `detach: false` would make sense, but it's more predictable to always default to false.

**Comment by denise-sanders** [comment]
(idk if this is the place for this question) It is really trivial to implement this for non-detached containers, but for detached containers you have to wait for the container to stop running. Is it bad form for me to create a thread that will wait for the container to die, and then call remove?

**Comment by chouseknecht** [comment]
Is this more solvable already with docker_service? With docker_service you can stop, start and scale the app. It consumes compose, so it automatically handles multiple instances of the same service. 

docker_container is predicated on _name_ being a required option, and the assumption (right or wrong) is that the container name uniquely identifies the container.

**Comment by denise-sanders** [comment]
@chouseknecht I am not sure what you mean- are you saying the user should run docker-service whenever they want to delete the container after execution?

**Comment by srstsavage** [comment]
@chouseknecht yeah, I can't imagine how `docker_service` would solve this (running a container and removing it/cleaning up when it's finished running).

@deniseirvin I'm afk and can't test, but I thought using `detach: true` would already wait for the container to finish running?

**Comment by chouseknecht** [comment]
@deniseirvin the docker_service module consumes compose, which means you can describe an entire app (one container or many) and manage it just as you would using the compose CLI. I'm suggesting using docker_service rather than docker_container.

@shane-axiom Yes, you can wait for the container to finish running via detach: false

Rather than adding yet another parameter to docker_container, I think a better way of handling this is with the a new state value of 'cleaned'. Here's the implementation: https://github.com/ansible/ansible-modules-core/pull/4144.

**Comment by srstsavage** [comment]
State value of `cleaned` is sensible to me. Thanks!

**Comment by denise-sanders** [comment]
I agree!

**Comment by chouseknecht** [comment]
Fixed with PR #4144. Ended up adding a `cleanup` parameter to be used in conjunction with `detach`. For example...

```
- docker_container:
    image: busybox
    command: sleep 15
    detach: false
    cleanup: yes
```
